contributors/design-proposals/node/node-fence.md

```diff
@@ -0,0 +1,528 @@
+
+Note: to keep those fence configuration we had two options - Adding attributes in metadata section to the existing node object or create ConfigMap that describe this information for each node with fencing options.
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+name: node_fence_config
+data: 
+  node_fence.properties: | 
+node_name=xxx
+isolation=[method-storage-iso1]]
+power_managment=[method-pm1, method-kdump]
+recovery=[method1,method2,method3]
+execution_policy=[seq\cun]
+```
+Notes:
+- Each configmap we relate a node name - this should be unique, and in the properties we set order for stages’ names.
+- execution_policy means: In list of method we might want to run all methods until all finish, or until only one finishes successfully (useful for sequential pm actions). 
+- Isolation\power_managment\recovery are parameters with list value which represent the method to run in each step of the fencing.
+#### Fence Method Config ###
+Represents the fence method parameters. This is done by configuring a “template” and specific config if needed with overloaded parameters.
+- Template is an abstract structure for a fence method (see example below for fence_ipmi device).
+- For each template the admin can create specific method config with different properties’ values.
+The template allows to reuse parameters for common devices and agents
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+ - name: fence_method_config
+data: 
+ - method.properties: | 
+node=node1_iscsi_pm_uenf324
+template=fence_method_template1
+template_param1_name=value1
+template_param2_name=value2
+..
+
+kind: ConfigMap
+apiVersion: 1.0
+metadata:
+ - name: fence_template_config
+ - version: 1.0
+data: 
+ - method_template.properties: |
+name=ipmilan_pm_device
+agent_name=fence_ipmilan
+must_success=Yes\No
+template_param1_name=value1
+template_param2_name=value2
+...
+```
+
+### Implementation
+#### Fence Controller
+The fence controller is a stateless pod instance running in cluster that supports fence. The controller contains two main loop: 1) identifies unresponsive nodes. 2) following fence crds.
+
+First controller will trigger one executor job to live in cluster.
+
+Identification: The controller will identify unresponsive node by check their apiserver object, once node becomes “not ready” a fence treatment will be triggered by posting crd for fence that includes all the information to execute fence stages by Fence Executor. 
+
+Following fence process using crd: The controller will check iteratively all fence objects and validate their execution (base on their status and state in stage). If execution is recognized as stuck or failed, the controller will handle the running job and run a new one on different resource to handle the request. Also, the controller is responsible to move to next “step” when node is still not ready, crd already exists, and the step is running for a configuration amount of time.
```

Comment:
Explain why you are using a CRD vs setting annotations / state on the node.

Reply:
Mainly to avoid adding more noise in the node object. previously in the crd I stored the list of methods to perform, now I changed it to hold only the "step" and the executor fetch the methods list. As it is now its reasonable to set it as additional annotations to the node object. I will test this option as well, it simplifies the solution of course.